### PR TITLE
[FSDP] Added sharded module 4/n

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Adds a `sharded_module` property to the lightning module and trainer. This is required to access the wrapped FSDP model when setting up optimizers which will be added in the FSDP PR. When not using sharded, this just uses the accelerator stored model.

Currently the only other option is to ask users to do `self.trainer.accelerator.training_type_plugin.model` within `configure_optimizers` which is cumbersome.

Related to #6152 

## Explanation

`sharded_module` really is: `wrapped_model` (i.e `DDP(lightning_module)` or `FSDP(lightning_module)`). The reason we need the property is because the model states are partitioned by FSDP and references to original weights are deleted.

For example, if you implemented your `configure_optimizers` like below:

```python
def configure_optimizers(self):
    return torch.optim.Adam(self.layer.parameters(), ...)
```

when FSDP wraps your model, it flattens all your weights down and destroys the original references, so self.layer.parameters() does not exist anymore (here is the issue to track: https://github.com/facebookresearch/fairscale/issues/430#issuecomment-785463057). This is currently a limitation of FSDP that will not be changed for the foreseeable future it seems. To remedy this:

1. configure optimizers after wrapping the model if the plugin requests it (Merged)
2. Introduce a way to access the internal wrapped model from the lightning module

Originally for 2., I did `self.trainer.accelerator.training_type_plugin.model` but this is very verbose for the user. I then made a property in the lightning module called `self.accelerator_wrapped_model` which points to `FSDP(lightning_module)` . I then went with `self.sharded_module` but from conversation with @justusschock on the subject, might be better to revert to a different name that's more general.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1105,6 +1105,14 @@ def call_configure_sharded_model(self, model: LightningModule) -> None:
             model.call_configure_sharded_model_hook = True
             self.accelerator.call_configure_sharded_model_hook = False
 
+    @property
+    def sharded_module(self) -> torch.nn.Module:
+        """
+        Returns: The Sharded version of the ``LightningModule`` module if using sharded,
+            else the model stored in the accelerator.
+        """
+        return self.accelerator.sharded_module
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1114]: Would it make sense to do an instance check here for the plugin and call sharded_module on the plugin? Instead of having a specific sharded_module property on the accelerator.

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -1465,6 +1465,18 @@ def tbptt_split_batch(self, batch, split_size):
 
         return splits
 
+    @property
+    def sharded_module(self) -> torch.nn.Module:
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1469]: Wouldn't it be better if this was grouped with the other properties?

Comment (SeanNaren) [reply pytorch_lightning/core/lightning.py:1469]: Yes it would, let me move it :)

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (SeanNaren) [reply pytorch_lightning/trainer/trainer.py]: This would mean the trainer is aware of the training type plugin, hence routing it through the accelerator

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 3]
```diff
@@ -284,3 +284,14 @@ def call_configure_sharded_model_hook(self) -> bool:
...
+    @property
+    def sharded_module(self) -> torch.nn.Module:
+        """
+        Override to pass a sharded version of the module. This is useful for configuring optimizers and training
+        with the sharded version of the ``LightningModule``.
+
+        Returns: The Sharded version of the ``LightningModule`` if using a Sharded Plugin,
+            else the model stored in the training type plugin.
+        """
+        return self.model
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:297]: Currently the default is `self.model`. In single device, this is just the lightning module. Using DDP, this is the `DistributedDataParallel` wrapper. In Sharded Training, this is the `ShardedDataParallel` wrapper. In Fully Sharded, this will be the `FullyShardedDataParallel` wrapper.

I went for the wrapper class rather than the lightning module for everything except FSDP because it standardizes across DDP/Sharded/Fully Sharded, and enforces the same logic (you can access the module via model.module, but be wary).

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:297]: isn't it misleading if the user code contains a call to `self.sharded_module` and the rest of the code is plugin/device agnostic?

**Comment by SeanNaren** [comment]
Moving this to draft, as the name of the variable is up for discussion. Suggestions so far are `self.accelerator_wrapped_model` `self.wrapped_model`

**Comment by SeanNaren** [comment]
So funnily enough, @awaelchli mentioned `self.trainer.model` already gives the wrapped model, which is the FSDP model. this can be used as a drop in replacement instead of the proposed changes in this PR, so closing this PR :)
